File: app/ReleaseImageGenerator.Domain/Implementations/ImageGenerator.cs
using ReleaseImageGenerator.Domain.Interfaces;
using SkiaSharp;
using Wacton.Unicolour;

namespace ReleaseImageGenerator.Domain.Implementations;

public class ImageGenerator : IImageGenerator
{
    public string? Text { get; set; }
    public int Width { get; set; }
    public int Height { get; set; }
    public SupportedFontFamily FontFamily { get; set; }
    public SupportedFontWeight FontWeight { get; set; }
    public string? PrimaryColor { get; set; }
    public SupportedImageFormat ImageFormat { get; set; }

    public ImageGenerator(ImageGeneratorOptions options)
    {
        Text = options.text;
        Width = options.width;
        Height = options.height;
        FontFamily = options.fontFamily;
        FontWeight = options.fontWeight;
        PrimaryColor = options.primaryColor;
        ImageFormat = options.imageFormat;
    }

    public MemoryStream GenerateImage()
    {
        using var surface = SKSurface.Create(new SKImageInfo(Width, Height));
        var canvas = surface.Canvas;
        var random = new Random();

        var primaryColor = PrimaryColor != null
            ? new Unicolour(PrimaryColor)
            : ColorGenerator.GetRandomColor(ColorGenerator.ColorLimitation.NEUTRAL_LIGHTNESS,
                ColorGenerator.ColorLimitation.NEUTRAL_SATURATION);
        BackgroundGenerator.GenerateBackground(canvas, Width, Height, random, primaryColor);
        PatternGenerator.GeneratePattern(canvas, Width, Height, random, primaryColor);
        NoiseGenerator.GenerateNoise(canvas, Width, Height, random);
        if (Text != null) TextGenerator.GenerateText(canvas, Text, Width, Height, FontFamily, FontWeight, primaryColor);

        // Encode and return image in the requested format
        var stream = new MemoryStream();
        using var image = surface.Snapshot();
        using var data = image.Encode(ImageFormat switch
        {
            SupportedImageFormat.jpeg => SKEncodedImageFormat.Jpeg,
            SupportedImageFormat.jpg => SKEncodedImageFormat.Jpeg,
            SupportedImageFormat.png => SKEncodedImageFormat.Png,
            SupportedImageFormat.webp => SKEncodedImageFormat.Webp,
            _ => SKEncodedImageFormat.Png
        }, 100);
        data.SaveTo(stream);
        return stream;
    }
}

File: src/lib/ImageGenerator.ts
import { createCanvas } from "canvas";
import type { ImageGeneratorOptions } from "./types";
import { SupportedImageFormat } from "./types";
import { ColorGenerator, ColorLimitation } from "./ColorGenerator";
import { BackgroundGenerator } from "./BackgroundGenerator";
import { PatternGenerator } from "./PatternGenerator";
import { NoiseGenerator } from "./NoiseGenerator";
import { TextGenerator } from "./TextGenerator";

export class ImageGenerator {
  private options: ImageGeneratorOptions;
  private randomSeed: number;

  constructor(options: ImageGeneratorOptions) {
    this.options = options;
    this.randomSeed = Math.abs(options.seed ?? Date.now());
  }

  // Seeded random number generator for consistent randomization
  private random = (): number => {
    this.randomSeed = (this.randomSeed * 9301 + 49297) % 233280;
    return this.randomSeed / 233280;
  };

  generateImage(): Buffer {
    const {
      width,
      height,
      text,
      fontFamily,
      fontWeight,
      primaryColor,
      imageFormat,
      patternType,
      noiseLevel,
    } = this.options;

    // Create canvas
    const canvas = createCanvas(width, height);
    const ctx = canvas.getContext("2d");

    // Get or generate primary color
    const color = primaryColor
      ? ColorGenerator.parseColor(this.random, primaryColor)
      : ColorGenerator.getRandomColor(
          this.random,
          ColorLimitation.NEUTRAL_LIGHTNESS,
          ColorLimitation.NEUTRAL_SATURATION,
        );

    // Generate layers
    BackgroundGenerator.generateBackground(
      ctx,
      width,
      height,
      this.random,
      color,
    );
    PatternGenerator.generatePattern(
      ctx,
      width,
      height,
      this.random,
      color,
      patternType,
    );
    NoiseGenerator.generateNoise(ctx, width, height, noiseLevel, this.random);

    if (text) {
      TextGenerator.generateText(
        ctx,
        text,
        width,
        height,
        fontFamily,
        fontWeight,
        color,
      );
    }

    // Encode to buffer
    let mimeType: string;
    switch (imageFormat) {
      case SupportedImageFormat.jpeg:
      case SupportedImageFormat.jpg:
        mimeType = "image/jpeg";
        break;
      case SupportedImageFormat.webp:
        mimeType = "image/webp";
        break;
      default:
        mimeType = "image/png";
    }

    return canvas.toBuffer(mimeType as any, { quality: 1.0 });
  }
}
